wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1199,18 +1201,60 @@ let init_vars (vars : Var.Set.t) (env : Env.t) : Constr.t list * Env.t =
         debug "Initializing var: %s\n%!" (Constr.to_string comp);
         comp :: inits, env)
 
+let user_func_spec (sub_name : string) (sub_pre : string) (sub_post : string)
+    (sub : Sub.t) (_ : Arch.t) : Env.fun_spec option =
+  debug "Making user-defined subroutine spec with subroutine-name: %s, pre:
+%s, post: %s \n%!" sub_name sub_pre sub_post;
+  if String.equal sub_name (Sub.name sub) then
+    Some {
+      spec_name = "user_func_spec";
+      spec = Summary (fun env post tid ->
+          (* turn strings into proper smtlib2 statements; incr stack_ptr *)
+          let sub_pre : Constr.t = Z3_utils.mk_smtlib2_single env sub_pre in
+          let sub_post : Constr.t = Z3_utils.mk_smtlib2_single env sub_post in
+          let sub_post, env = increment_stack_ptr sub_post env in
+          (* collect (physical) inputs/outputs of sub *)
+          let sub_inputs : Var.t list = get_vars env sub |> Var.Set.to_list in
+          let sub_inputs : Var.t list =
+            List.filter sub_inputs ~f:(fun v -> Var.is_physical v) in
+          let sub_outputs : Var.t list = sub_inputs in
+          (* replace init_vars with vars inside sub_post *) 
```

Comment:
???

Reply:
I moved the comment `(* replace init_vars with vars inside sub_post *)` to right before where the replacement is made, namely `let sub_post = Constr.substitute sub_post inits regs in`

Reply:
Great.